pep-0697.rst

```diff
@@ -0,0 +1,357 @@
+PEP: 697
+Title: C API for Extending Opaque Types
+Author: Petr Viktorin <encukou@gmail.com>
+Status: Draft
+Type: Standards Track
+Content-Type: text/x-rst
+Created: 23-Aug-2022
+Python-Version: 3.12
+
+
+Abstract
+========
+
+Add limited C API for extending types whose ``struct`` is opaque,
+by allowing code to only deal with data specific to a particular (sub)class.
+
+Make the mechanism usable with ``PyHeapType``.
+
+
+Motivation
+==========
+
+Extending opaque types
+----------------------
+
+In order to allow changing/optimizing CPython, and allow freedom for alternate
+implementations of the C API, best practice is to not expose memory layout
+(C structs) in public API, and instead rely on accessor functions.
+(When this hurts performance, direct struct access can be allowed in a
+less stable API tier, at the expense of compatibility with diferent
+versions/implementations of the interpreter.)
+
+However, when a particular type's instance struct is hidden, it becomes
+difficult to subclass it.
+The usual subclassing pattern, explained `in the tutorial <https://docs.python.org/3/extending/newtypes_tutorial.html#subclassing-other-types>`_,
```

Comment:
Minor sidenote: You could consider using an intersphinx reference for these cases rather than a hardcoded URL—though you'll need to add proper ref target labels to the source tutorial headings.

You can also use it to easily and reliably link existing functions, structs, etc., if you think it prudent, instead of hardcoding links or making the reader search for them themselves if they want more details.

Reply:
BTW, see https://github.com/python/peps/pull/2773#issuecomment-1229104900

Reply:
So, I'm trying to link to [Python 3.11's `PyListObject`](https://docs.python.org/3.11/c-api/list.html?highlight=pylistobject#c.PyListObject). I added a `python3.11` mapping and tried:
```
:c:type:`python3.11:PyListObject`
```
but I'm getting:
```
WARNING: Unparseable C cross-reference: 'python3.11:PyListObject'
Invalid C declaration: Expected identifier in nested name. [error at 8]
  python3.11:PyListObject
  --------^
```
So I read Intersphinx docs, and got this to work:
```
:external+python3.11:c:type:`PyListObject`
```
Is that the way? If so, it doesn't sound like something to burden PEP authors with, tbh